synthesizer/src/vm/mod.rs

```diff
@@ -335,13 +335,34 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         // Attention: The following order is crucial because if 'finalize' fails, we can rollback the block.
         // If one first calls 'finalize', then calls 'insert(block)' and it fails, there is no way to rollback 'finalize'.
 
+        // Enable the atomic batch override, so that both the insertion and finalization belong to a single batch.
+        #[cfg(feature = "rocks")]
+        assert!(self.block_store().flip_atomic_override()?);
+
         // First, insert the block.
         self.block_store().insert(block)?;
         // Next, finalize the transactions.
         match self.finalize(state, block.ratifications(), block.solutions(), block.transactions()) {
-            Ok(_ratified_finalize_operations) => Ok(()),
+            Ok(_ratified_finalize_operations) => {
+                // Disable the atomic batch override, executing it.
+                #[cfg(feature = "rocks")]
+                assert!(!self.block_store().flip_atomic_override()?);
+                Ok(())
+            }
             Err(finalize_error) => {
-                // Rollback the block.
+                // Clear all the pending atomic operations so that disabling the atomic batch override
+                // doesn't execute any storage operation that has accumulated since it was enabled.
+                // This only applies to persistent storage - the in-memory one needs to roll back the
+                // latest block.
+                #[cfg(feature = "rocks")]
+                self.block_store().atomic_abort();
+                #[cfg(feature = "rocks")]
+                self.finalize_store().atomic_abort();
+                // Disable the atomic batch override.
+                #[cfg(feature = "rocks")]
+                assert!(!self.block_store().flip_atomic_override()?);
+                // Rollback the block (only applies to in-memory storage).
+                #[cfg(not(feature = "rocks"))]
```

Comment:
This is not sufficient.

`self.block_store().insert(block)?;` updates the in-memory merkle tree. However since we are not calling `self.block_store().remove_last_n(1)` anymore in the rocksdb case, then the tree will not be correct for the given db state.

Reply:
good catch :+1:, I reverted this bit; the `remove_last_n` code will still attempt to remove the non-existent blocks in case of persistent storage, but that shouldn't be noticable